pgdrive/envs/marl_envs/marl_intersection.py

```diff
@@ -0,0 +1,336 @@
+from pgdrive.scene_creator.road.road import Road
+from pgdrive.utils import get_np_random, norm, PGConfig
+
+MAIntersectionConfig = dict(
+    map_config=dict(exit_length=60, lane_num=2),
+    top_down_camera_initial_x=95,
+    top_down_camera_initial_y=15,
+    top_down_camera_initial_z=120
+)
+
+
+class MAIntersectionMap(PGMap):
+    def _generate(self, pg_world):
+        length = self.config["exit_length"]
+
+        parent_node_path, pg_physics_world = pg_world.worldNP, pg_world.physics_world
+        assert len(self.road_network.graph) == 0, "These Map is not empty, please create a new map to read config"
+
+        # Build a first-block
+        last_block = FirstBlock(
+            self.road_network,
+            self.config[self.LANE_WIDTH],
+            self.config[self.LANE_NUM],
+            parent_node_path,
+            pg_physics_world,
+            1,
+            length=length
+        )
+        self.blocks.append(last_block)
+
+        # Build Intersection
+        InterSection.EXIT_PART_LENGTH = length
+        last_block = InterSection(1, last_block.get_socket(index=0), self.road_network, random_seed=1)
+        last_block.construct_block(
+            parent_node_path,
+            pg_physics_world,
+            extra_config={
+                "exit_radius": 10,
+                "inner_radius": 30,
+                "angle": 70,
+                # Note: lane_num is set in config.map_config.lane_num
+            }
+        )
+        self.blocks.append(last_block)
+
+
+class MultiAgentIntersectionEnv(MultiAgentPGDrive):
+    spawn_roads = [
+        Road(FirstBlock.NODE_2, FirstBlock.NODE_3),
+        -Road(InterSection.node(1, 0, 0), InterSection.node(1, 0, 1)),
+        -Road(InterSection.node(1, 1, 0), InterSection.node(1, 1, 1)),
+        -Road(InterSection.node(1, 2, 0), InterSection.node(1, 2, 1)),
+    ]
+
+    @staticmethod
+    def default_config() -> PGConfig:
+        return MultiAgentPGDrive.default_config().update(MAIntersectionConfig, allow_overwrite=True)
+
+    def _update_map(self, episode_data: dict = None, force_seed=None):
+        if episode_data is not None:
+            raise ValueError()
+        map_config = self.config["map_config"]
+        map_config.update({"seed": self.current_seed})
+
+        if self.current_map is None:
+            self.current_seed = 0
+            new_map = MAIntersectionMap(self.pg_world, map_config)
+            self.maps[self.current_seed] = new_map
+            self.current_map = self.maps[self.current_seed]
+
+    def _update_destination_for(self, vehicle):
+        # when agent re-joined to the game, call this to set the new route to destination
+        end_roads = copy.deepcopy(self.spawn_roads)
+        end_roads.remove(vehicle.current_road)
+        end_road = -get_np_random(self._DEBUG_RANDOM_SEED).choice(end_roads)  # Use negative road!
+        vehicle.routing_localization.set_route(vehicle.lane_index[0], end_road.end_node)
```

Comment:
Why this function is different from MARound?

Reply:
No circular route to drive to reverse road in the intersection block

Reply:
The modification in set_route function is also used to solve this problem.

Reply:
Why not~~

Reply:
We should allow U-turn too!